### PR TITLE
Bytevector/Vector support in Data, parser

### PR DESCRIPTION
**Pull request by CinchBlue**
I've revised Bytevector and Vector code, and integrated it into `shaka::Data` and the parsing code. There are new test cases and small revisions here and there.

**Please test whether the parsing rules and bytevector/vector creation code works.** The REPL should automatically have support for bytevectors and vectors since it uses the same parsing logic as the unit/integration tests.

**Please test whether bytevector/vector can be used successfully from `shaka::Data`.**

### CODE REVIEW

File: src/shaka_scheme/system/core/lists.cpp [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+namespace shaka {
+namespace core {
+
+NodePtr cons(NodePtr left, NodePtr right) {
+  return create_node(Data(DataPair(left, right)));
+}
+
+NodePtr car(NodePtr node) {
+  if (node->get_type() != Data::Type::DATA_PAIR) {
+    throw shaka::TypeException(10001, "car(): Data does not hold DataPair");
```
Comment (Kayla-Kwock) [inline src/shaka_scheme/system/core/lists.cpp:12]: You're already using namespace `shaka`. Unless i'm misunderstanding something, the `shaka::` is unnecessary.

File: src/shaka_scheme/system/core/lists.cpp [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+
+bool is_null_list(NodePtr node) {
+  return node->get_type() == Data::Type::NULL_LIST;
+}
+
+bool is_proper_list(NodePtr node) {
+  // The empty list is a proper list
+  if (is_null_list(node)) { return true; }
+  // Lists must be pairs.
+  if (is_null_list(node)) { return true; }
```
Comment (Kayla-Kwock) [inline src/shaka_scheme/system/core/lists.cpp]: Repeated `if` case

File: src/shaka_scheme/system/core/vectors.cpp [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+/**
+ * @brief Converts a proper list into a vector.
+ * @param list A proper list, terminated with a null list
+ * @return A vector containing the same elements as the original list
+ */
+NodePtr list_to_vector(NodePtr list) {
+  // Find the length of the list.
+  std::size_t length = core::length(list);
+  shaka::Vector vector(length);
```
Comment (Kayla-Kwock) [inline src/shaka_scheme/system/core/vectors.cpp]: Already using namespace `shaka` and namespace `core`

File: tst/shaka_scheme/system/core/unit-core-vectors.cpp [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  // Then: the elements should be #(1 2 3)
+  ASSERT_EQ(v[0]->get<Number>(), Number(1));
+  ASSERT_EQ(v[1]->get<Number>(), Number(2));
+  ASSERT_EQ(v[2]->get<Number>(), Number(3));
+}
+
+/**
+ * @brief Test: Testing (list->vector l)
+ */
+TEST(VectorsUnitTest, vector_to_list_2_elements) {
```
Comment (Kayla-Kwock) [inline tst/shaka_scheme/system/core/unit-core-vectors.cpp]: Shouldn't this be `vector_to_list_3_elements` since you created a vector containing 3 numbers

**Review by Kayla-Kwock — COMMENTED** [review]
Looked over Vector and Bytevector changes regarding implementation and test cases. Parser looks good, but would prefer if someone else double checks it.

File: src/shaka_scheme/system/core/lists.cpp (hunk 1 above)
Comment (CinchBlue) [reply src/shaka_scheme/system/core/lists.cpp:12]: That's true. Just being explicit here that it's a `shaka::TypeException` and not some other `TypeException` class in the future.

File: src/shaka_scheme/system/core/lists.cpp (hunk 2 above)
Comment (CinchBlue) [reply src/shaka_scheme/system/core/lists.cpp]: That's a good catch. Thanks!

File: tst/shaka_scheme/system/core/unit-core-vectors.cpp (hunk 4 above)
Comment (CinchBlue) [reply tst/shaka_scheme/system/core/unit-core-vectors.cpp]: That's true. Let me fix that.

File: src/shaka_scheme/system/core/vectors.cpp (hunk 3 above)
Comment (CinchBlue) [reply src/shaka_scheme/system/core/vectors.cpp]: `core::length` is unambiguous but also specific enough such that it's different from the slew of `length` functions out there. I can change this to `shaka::core::length` to make it more explicit.

**Comment by CinchBlue** [comment]
I have made some edits as suggested by @Kayla-Kwock to the pull request branch.

- Fixed the typo in `core/lists.cpp`.
- Made the namespacing more clear in the other two source files.

**Review by lihongman — APPROVED** [review]
I do not see any problems (there might be but I just did not catch them). Otherwise, it seems good to go.

**Pull request merged**